Fix part of #10415: Refactor activity validators

hardikkat24 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #10415 .
2. This PR does the following: Refactors activity validators to a separate file. This separates `exploration` related validators. 

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes made in this PR work correctly.
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

hardikkat24 committed:
added exploration_validators.py

hardikkat24 committed:
added exploration_validators_test.py

hardikkat24 committed:
documentation improved

hardikkat24 committed:
documentation improved-2

hardikkat24 committed:
prod_validators unused imports removed

hardikkat24 committed:
prod_validators_test unused imports removed

hardikkat24 committed:
changes

hardikkat24 committed:
Merge branch 'develop' into issue-10415

hardikkat24 commented:
@vojtechjelinek @aks681 PTAL
Why are the backend tests failing?

vojtechjelinek commented:
@hardikkat24 Please take a look at https://groups.google.com/g/oppia-dev/c/k6ppgwvq3hI

hardikkat24 committed:
Add new validator test to shards

hardikkat24 commented:
@vojtechjelinek Thanks! I have added `exploration_validators_test` to shards.

hardikkat24 committed:
Merge branch 'develop' into issue-10415

hardikkat24 commented:
@vojtechjelinek Why are the tests failing? I can't figure out.

sajalasati commented:
> @vojtechjelinek Why are the tests failing? I can't figure out.

@hardikkat24 Can you try updating with develop once and push changes again?

hardikkat24 committed:
Merge branch 'develop' into issue-10415

hardikkat24 commented:
@sajalasati Done!

vojtechjelinek commented:
@hardikkat24 Look into feedback_validators_test.py some of the tests there are failing, maybe because of some constant that you have removed from prod_validators.py and that are used from feedback_validators.py

hardikkat24 commented:
@vojtechjelinek Thanks a lot!
I removed `TARGET_TYPE_TO_TARGET_MODEL` from `prod_validators.py` which was causing this problem.
Should I add the constants I removed from `prod_validators.py` because other files may also be importing from `prod_validators`.
But our plan is to remove `prod_validators` right?

PS: I have searched now, I think only: `core/domain/feedback_jobs_one_off.py`, `core/domain/feedback_validators.py`, `core/domain/feedback_validators_test.py` import `prod_validators`

vojtechjelinek commented:
@hardikkat24 If the `TARGET_TYPE_TO_TARGET_MODEL` is used in all these files then feel free to move it to feconf.py. In feconf.py there is `SUGGESTION_TARGET_TYPE_CHOICES` that you can replace with `TARGET_TYPE_TO_TARGET_MODEL` and then in the places where `SUGGESTION_TARGET_TYPE_CHOICES` was used use `TARGET_TYPE_TO_TARGET_MODEL.keys()`.

hardikkat24 commented:
@vojtechjelinek Thanks a lot!
I did as you said but I think these lines in `feconf` are giving errors.
```
(
    exp_models, question_models,
    skill_models, topic_models
) = models.Registry.import_models([
    models.NAMES.exploration, models.NAMES.question,
    models.NAMES.skill, models.NAMES.topic
])
```

The error is:
```
Traceback (most recent call last):
  File ".git/hooks/pre-commit", line 40, in <module>
    from scripts import common  # isort:skip # pylint: disable=wrong-import-position
  File "/home/hardik/opensource/oppia/scripts/common.py", line 33, in <module>
    import feconf
  File "/home/hardik/opensource/oppia/feconf.py", line 35, in <module>
    models.NAMES.skill, models.NAMES.topic
  File "/home/hardik/opensource/oppia/core/platform/models.py", line 336, in import_models
    return cls._get().import_models(model_names)
  File "/home/hardik/opensource/oppia/core/platform/models.py", line 108, in import_models
    from core.storage.exploration import gae_models as exp_models
  File "/home/hardik/opensource/oppia/core/storage/exploration/gae_models.py", line 26, in <module>
    import core.storage.base_model.gae_models as base_models
  File "/home/hardik/opensource/oppia/core/storage/base_model/gae_models.py", line 28, in <module>
    transaction_services = models.Registry.import_transaction_services()
  File "/home/hardik/opensource/oppia/core/platform/models.py", line 396, in import_transaction_services
    return cls._get().import_transaction_services()
  File "/home/hardik/opensource/oppia/core/platform/models.py", line 209, in import_transaction_services
    from core.platform.transactions import gae_transaction_services
  File "/home/hardik/opensource/oppia/core/platform/transactions/gae_transaction_services.py", line 24, in <module>
    from google.appengine.ext import ndb
ImportError: No module named appengine.ext
```
This error is thrown when I tried to commit my changes.
Can you explain how should I proceed?
Thanks!

vojtechjelinek commented:
@hardikkat24 This seems to be some issue with your local installation. Can you try running `python -m scripts.install_third_party_libs`?

hardikkat24 commented:
@vojtechjelinek I tried running this. But this shows same error.
However when I commented out the following code snippet, everything was working fine.
```
(
    exp_models, question_models,
    skill_models, topic_models
) = models.Registry.import_models([
    models.NAMES.exploration, models.NAMES.question,
    models.NAMES.skill, models.NAMES.topic
])
```
Even the commit checks uses `feconf.py`. So do you think this error is due to this code snippet. 
I think what we can do is that we can initialize `TARGET_TYPE_TO_TARGET_MODEL` in `feedback_validators` and import it in other two files. And we can leave `feconf.py` unchanged. What do you think?

PS: I tried adding some imports in `feconf.py` such as `from core.domain import base_model_validators`, this also shows a different error when I try to run the server or when I try to commit.
Can you add such imports in your `feconf.py` and run server without encountering an error?

vojtechjelinek commented:
> @vojtechjelinek I tried running this. But this shows same error.
> However when I commented out the following code snippet, everything was working fine.
> 
> ```
> (
>     exp_models, question_models,
>     skill_models, topic_models
> ) = models.Registry.import_models([
>     models.NAMES.exploration, models.NAMES.question,
>     models.NAMES.skill, models.NAMES.topic
> ])
> ```
> 
> Even the commit checks uses `feconf.py`. So do you think this error is due to this code snippet.
> I think what we can do is that we can initialize `TARGET_TYPE_TO_TARGET_MODEL` in `feedback_validators` and import it in other two files. And we can leave `feconf.py` unchanged. What do you think?
> 
> PS: I tried adding some imports in `feconf.py` such as `from core.domain import base_model_validators`, this also shows a different error when I try to run the server or when I try to commit.
> Can you add such imports in your `feconf.py` and run server without encountering an error?

No, do not import anything in feconf.py. Keep the changes as you ahve them and try committing and pushing with using `--no-verify` flag that should not fail (I will take a look at the code then to see if there is some mistake). I still think the error is caused locally only for you.

hardikkat24 commented:
@vojtechjelinek PTAL, the same errors are thrown by checks. Thanks!

vojtechjelinek commented:
@hardikkat24 Ohhh, right, you cannot import models in feconf.py. Sorry for the confusion. I think the best place to put `TARGET_TYPE_TO_TARGET_MODEL` would be the feedback_services.py and use it from there.

hardikkat24 committed:
added TARGET_TO_MODEL to feedback_services

hardikkat24 committed:
Merge branch 'develop' into issue-10415

hardikkat24 committed:
fixed bugs

hardikkat24 commented:
@vojtechjelinek Thanks a lot! I have made the changes. PTAL.
I wanted to ask one thing, why can't we import models in `feconf.py`?
I can't figure out why the tests are failing?

vojtechjelinek commented:
> @vojtechjelinek Thanks a lot! I have made the changes. PTAL.
> I wanted to ask one thing, why can't we import models in `feconf.py`?
> I can't figure out why the tests are failing?

Since feconf.py is ought to be used as a constants file and it is imported in many places in our code. Adding imports into it leverages the risk of circular imports.

The two tests that are failing now are just flaky and only need to be restarted.

vojtechjelinek reviewed (approved):
LGTM!

U8NWXD commented on scripts/backend_test_shards.json:
```diff
@@ -192,7 +192,8 @@
         "core.domain.user_jobs_one_off_test",
-        "core.domain.topic_validators_test"
+        "core.domain.topic_validators_test",
+        "core.domain.exploration_validators_test"
```
With this addition, shard 9 is taking 33 mins to run, which is longer than our goal of 30 mins. Could you add this to shard 4 instead, which currently takes only 13 mins?

![Screen Shot 2021-02-16 at 21 30 46](https://user-images.githubusercontent.com/19878639/108148213-5d74a000-709e-11eb-81d3-0ebb95ef7fe3.png)

In general, you should add new tests to the shard with the fastest run-time. This guidance is on the [backend tests wiki page](https://github.com/oppia/oppia/wiki/Writing-backend-tests#guidelines-for-writing-good-tests) in case you forget.

hardikkat24 replied:
@U8NWXD Thanks! Will keep this in mind. I have made changes.

hardikkat24 replied:
Done

hardikkat24 committed:
test moved to shard 4

hardikkat24 commented:
@U8NWXD PTAL

hardikkat24 committed:
Merge branch 'develop' into issue-10415

sajalasati commented:
@aks681 PTAL!

hardikkat24 committed:
Merge branch 'develop' into issue-10415

hardikkat24 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into issue-10415

hardikkat24 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into issue-10415

aks681 reviewed (approved):
Lgtm as codeowner.

vojtechjelinek committed:
Merge branch 'develop' into issue-10415

vojtechjelinek committed:
Merge branch 'develop' into issue-10415
